storage/innobase/include/fut0lst.h

```diff
@@ -117,11 +117,8 @@ inline uint32_t flst_get_len(const flst_base_node_t *base)
 /** @return a file address */
 inline fil_addr_t flst_read_addr(const byte *faddr)
 {
-  fil_addr_t addr= { mach_read_from_4(faddr + FIL_ADDR_PAGE),
-		     mach_read_from_2(faddr + FIL_ADDR_BYTE) };
-  ut_a(addr.page == FIL_NULL || addr.boffset >= FIL_PAGE_DATA);
-  ut_a(ut_align_offset(faddr, srv_page_size) >= FIL_PAGE_DATA);
```

Comment:
You are right, we could keep it, possibly demoted to `ut_ad()` in order to avoid unnecessary bloat in release builds. But, while I worked on this, I ensured that the `ut_align_offset()` assertion should actually be satisfied everywhere. Anything that would write data includes a debug assertion that the data pointer is within the specified buffer pool page frame.

For example, in most calls via `flst_get_first()` or `flst_get_last()` you can see that the assertion is trivially satisifed, because we are adding at least (`FIL_PAGE_DATA` (such as `FSP_HEADER_OFFSET` and a little more) to the page frame address. In `fseg_alloc_free_extent()` before we invoke this on `inode + FSEG_FREE` we had checked that the `page_offset(inode)` is reasonable and flagged corruption if it was not. I think that it is better to have stricter checks in the callers.

Reply:
ffc0009aa6a9946e59bbcf70b2cd03a2a10cc424 replaces the assertion on `faddr` with a debug assertion. In each caller, there are stricter replacements of the assertion on `addr.boffset`.